File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        int[] arr = new int[nums.length];
        arr[0] = nums[0];
        int ret = 1;
        for(int it : nums){
            if(it > arr[ret - 1]){
                arr[ret++] = it; 
                continue;
            }
            int l = 0 , r = ret - 1 , mid = 0;
            while(l != r){
                mid = (l + r) / 2;
                if(it <= arr[mid]){
                    r = mid;
                }else{
                    l = mid + 1;
                }
            }
            arr[l] = it;
            if(ret == l){
                ++ret;
            }
        }
        return ret;
    }
}

File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int>v;
        int n = nums.size();
        int ret = 1;
        v.push_back(nums[0]);
        for(int i = 1 ; i < n ; i++){
            if(nums[i] > v.back()){
                ++ret;
                v.push_back(nums[i]);
            }else{
                v[lower_bound(v.begin() , v.end() , nums[i]) - v.begin()] = nums[i];
            }
        }
        return ret;
    }
};